baseplate/context/memcache/lib.py

```diff
@@ -0,0 +1,226 @@
+import logging
+import json
+import zlib
+
+from ..._compat import (
+    long,
+    pickle,
+    string_types,
+)
+
+
+"""Memcache serialization/deserialization (and compression) helper methods.
+
+Memcached can only store strings, so to store arbitrary objects we need to
+serialize them to strings and be able to deserialize them back to their
+original form.
+
+New services should use dump_and_compress() and decompress_and_load().
+
+Services that need to read and write to the same memcache instances as r2
+should use pickle_and_compress() and decompress_and_unpickle().
+
+"""
+
+
+class Flags(object):
+    """Memcached client flags
+
+    Flags are an arbitrary 16-bit unsigned integer that the memcache server
+    stores along with the data and sends back when the item is retrieved.
+    Clients may use this as a bit field to store data-specific information;
+    this field is opaque to the server.
+
+    """
+
+    JSON = 1 << 0
+    INTEGER = 1 << 1
+    LONG = 1 << 2
+    ZLIB = 1 << 3
+
+
+def decompress_and_load(key, serialized, flags):
+    """Deserialization method compatible with dump_and_compress().
+
+    :param str key: the memcached key.
+    :param str serialized: the serialized object returned from memcached.
+    :param int flags: value stored and returned from memcached for the client
+        to use to indicate how the value was serialized.
+    :returns str value: the deserialized value.
+
+    """
+
+    if flags & Flags.ZLIB:
+        serialized = zlib.decompress(serialized)
+        flags ^= Flags.ZLIB
+
+    if flags == 0:
+        return serialized
+    elif flags == Flags.INTEGER:
+        # python3 doesn't have a long integer type, so all integers are written
+        # with Flags.INTEGER. This means that a value written by a python3
+        # client can exceed the maximum integer value (sys.maxint). This
+        # appears to be ok--python2 will automatically convert to long if the
+        # value is too large.
+        return int(serialized)
+    elif flags == Flags.LONG:
+        return long(serialized)
+    else:
```

Comment:
maybe this should be `elif flags == Flags.JSON`, just in case, and the else clause could either throw an error, or log error and return serialized?